javalib/src/main/scala/java/util/Properties.scala

```diff
@@ -67,19 +101,325 @@ class Properties(protected val defaults: Properties)
   }
 
   def list(out: PrintStream): Unit = {
-    out.println("-- listing properties --")
-    entrySet().asScala.foreach { entry => out.println(format(entry)) }
+    list(new PrintWriter(new OutputStreamWriter(out), true))
   }
 
   def list(out: PrintWriter): Unit = {
     out.println("-- listing properties --")
     entrySet().asScala.foreach { entry => out.println(format(entry)) }
   }
 
+  def store(out: OutputStream, comments: String): Unit = {
+    if (out == null) {
+      throw new NullPointerException()
+    }
+    // https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.InputStream-
+    val writer = new OutputStreamWriter(out, "ISO8859_1")
+    store(writer, comments)
+  }
+
+  def store(writer: Writer, comments: String): Unit = {
+    if (writer == null) {
+      throw new NullPointerException()
+    }
+
+    if (comments != null) {
+      writeComments(writer, comments)
+    }
+
+    writer.write('#')
+    writer.write(new Date().toString)
+    writer.write(System.lineSeparator)
+
+    val buffer = new StringBuilder(200)
+    entrySet().asScala.foreach { entry =>
+      val key = entry.getKey.asInstanceOf[String]
+      dumpString(buffer, key, true, false)
+      buffer.append('=')
+      dumpString(buffer, entry.getValue.asInstanceOf[String], false, false)
+      buffer.append(System.lineSeparator)
+      writer.write(buffer.toString)
+      buffer.setLength(0)
+    }
+    writer.flush()
+  }
+
+  @deprecated("", "") def save(out: OutputStream, comments: String): Unit =
+    store(out, comments)
+
+  private val NONE     = 0
+  private val SLASH    = 1
+  private val UNICODE  = 2
+  private val CONTINUE = 3 // when \r is encountered looks for next \n
+  private val KEY_DONE = 4
+  private val IGNORE   = 5
+  private lazy val nextCharMap =
+    SMap('b' -> '\b', 'f' -> '\f', 'n' -> '\n', 'r' -> '\r', 't' -> '\t')
+
+  private def isEbcdic(in: BufferedInputStream): Boolean = {
+    var b: Byte = 0
+    while ({ b = in.read.toByte; b != -1 }) {
+      if (b == 0x23 || b == 0x0a || b == 0x3d) { // ascii: newline/#/=
+        return false
+      }
+      if (b == 0x15) { // EBCDIC newline
+        return true
+      }
+    }
+    //we found no ascii newline, '#', neither '=', relative safe to consider it
+    //as non-ascii, the only exception will be a single line with only key(no value and '=')
+    //in this case, it should be no harm to read it in default charset
+    false
+  }
+
+  private def loadImpl(reader: Reader): Unit = {
+    var mode           = NONE
+    var unicode        = 0
+    var count          = 0
+    var nextChar: Char = 0
+    var buf            = new Array[Char](80)
+    var offset         = 0
+    var keyLength      = -1
+    val br             = new BufferedReader(reader)
+
+    @tailrec def processNext(isFirstChar: Boolean): Unit = {
```

Comment:
Please avoid pseudo-tailrec functions that actually modify surrounding state. That's twice as hard to understand, because one has to reason about the state *plus* the accumulators like `isFirstChar`. Prefer using a good old `while` loop instead.

Reply:
Overall, I find this algorithm that revolves around reading one character per loop iteration, and fiddle with a `mode` to save the state, *very* hard to reason about.

Have you considered using a much more straightforward approach that parses logical bits instead? Something with the following structure:
```scala
while (...) {
  val line = br.readLine()
  var i: Int = 0

  def parseUnicodeEscape(): Char = ...

  def parseKey(): String = {
    while (line.charAt(i) != '=') { ... }
  }

  def parseValue(): String = {
    while (i < line.length) { ... }
  }

  val key = parseKey()
  i += 1 // the '='
  val value = parseValue()

  putProperty(key, value)
}
```
Wouldn't that be so much easier to reason about?

Reply:
I didn't write any of the current code but I knew it was super ugly. I couldn't reason about any of it honestly.

Thanks for the basic layout - I have made some progress. If I can get it to the point the tests pass then that would be great.

Reply:
I think the result is much more readable and maintainable. It has a very imperative style using `var`s and such.